### PR TITLE
Remove dependency on nebulous in proof generation

### PR DESCRIPTION
**Pull request by rahulghangas**
Use the same recursive approach as in nmt.Root() to generate proof
Removes the dependency on Celestiaorg/merkletree in proof generation

- [x] Passing full testing
- [x] Fixes https://github.com/celestiaorg/nmt/issues/15

### CODE REVIEW

**Review by evan-forbes — COMMENTED** [review]
LGTM, 

[blocking question] should we delete the entire internal package now that we're not using it?

**Review by liamsi — CHANGES_REQUESTED** [review]
Agree with @evan-forbes we should delete the internal package entirely.

File: proof_test.go [hunk 1]
```diff
@@ -3,14 +3,58 @@ package nmt
 import (
 	"bytes"
 	"crypto/sha256"
+	"io"
 	"testing"
 
 	"github.com/celestiaorg/merkletree"
```
Comment (liamsi) [inline proof_test.go:9]: Ideally, we would remove this dependency entirely. But orthogonal to this PR.

Comment (rahulghangas) [reply proof_test.go:9]: Done

**Comment by rahulghangas** [comment]
Whoops, keep on forgetting you can only request one review. Sorry

File: proof_test.go (hunk 1 above)
Comment (rahulghangas) [reply proof_test.go:9]: Edit -- This cannot be removed, merkletree relies on io errors internally, which is wild!

Comment (rahulghangas) [reply proof_test.go:9]: Anyway, this is only in a test, I think should be fine?

**Review by evan-forbes — APPROVED** [review]
LGTM

**Comment by adlerjohn** [comment]
Poking @liamsi for re-review since changes requested.

File: nmt.go [hunk 2]
```diff
@@ -127,12 +126,11 @@ func (n NamespacedMerkleTree) Prove(index int) (Proof, error) {
 func (n NamespacedMerkleTree) ProveRange(start, end int) (Proof, error) {
 	isMaxNsIgnored := n.treeHasher.IsMaxNamespaceIDIgnored()
 	n.computeLeafHashesIfNecessary()
-	subTreeHasher := internal.NewCachedSubtreeHasher(n.leafHashes, n.treeHasher)
 	// TODO: store nodes and re-use the hashes instead recomputing parts of the tree here
-	proof, err := merkletree.BuildRangeProof(start, end, subTreeHasher)
-	if err != nil {
-		return NewEmptyRangeProof(isMaxNsIgnored), err
+	if start < 0 || start > end || start == end || end > len(n.leafHashes) {
```
Comment (adlerjohn) [inline nmt.go:130]: Elsewhere in the code you use `>=`, so for consistency

```suggestion
	if start < 0 || start >= end || end > len(n.leafHashes) {
```

File: nmt.go [hunk 3]
```diff
@@ -171,27 +169,66 @@ func (n NamespacedMerkleTree) ProveNamespace(nID namespace.ID) (Proof, error) {
...
+func (n NamespacedMerkleTree) buildRangeProof(proofStart, proofEnd int) [][]byte {
+	proof := [][]byte{}
+	var recurse func(start, end int, includeNode bool) []byte
+	recurse = func(start, end int, includeNode bool) []byte {
+		if start >= len(n.leafHashes) {
+			return nil
+		}
+
+		// reached a leaf
+		if end-start == 1 {
```
Comment (adlerjohn) [inline nmt.go:189]: Did you run gofmt on this file? Doesn't look like we have it in CI, but this line looks sus.

Comment (rahulghangas) [reply nmt.go:189]: Actually, gofmt changes it to this form

**Comment by rahulghangas** [comment]
bulld is failing due to codecov upload failure

**Review by adlerjohn — APPROVED** [review]
utACK

**Comment by adlerjohn** [comment]
Poking @liamsi for re-review since changes requested.

**Comment by rahulghangas** [comment]
@liamsi bump

**Comment by rahulghangas** [comment]
@adlerjohn can be merged now?

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**